docs: msq autocompaction (#16681)
apache/druid: docs/data-management/automatic-compaction.md

:::info
 Auto-compaction skips datasources that have a segment granularity of `ALL`.
:::

As a best practice, you should set up auto-compaction for all Druid datasources. You can run compaction tasks manually for cases where you want to allocate more system resources. For example, you may choose to run multiple compaction tasks in parallel to compact an existing datasource for the first time. See [Compaction](compaction.md) for additional details and use cases.

This topic guides you through setting up automatic compaction for your Druid cluster. See the [examples](#examples) for common use cases for automatic compaction.

## Coordinator-based

The Coordinator [indexing period](../configuration/index.md#coordinator-operation), `druid.coordinator.period.indexingPeriod`, controls the frequency of compaction tasks.
The default indexing period is 30 minutes, meaning that the Coordinator first checks for segments to compact at most 30 minutes from when auto-compaction is enabled.

> This time period affects other Coordinator duties including merge and conversion tasks.

Reviewer: ```suggestion
This time period also affects other Coordinator duties such as cleanup of unused segments and stale pending segments.
```

Merge and conversion tasks do not exist in Druid anymore.

To configure the auto-compaction time period without interfering with `indexingPeriod`, see [Set frequency of compaction runs](#compaction-frequency).

At every invocation of auto-compaction, the Coordinator initiates a [segment search](../design/coordinator.md#segment-search-policy-in-automatic-compaction) to determine eligible segments to compact.
When there are eligible segments to compact, the Coordinator issues compaction tasks based on available worker capacity.
If a compaction task takes longer than the indexing period, the Coordinator waits for it to finish before resuming the period for segment search.

No additional configuration is needed to run automatic compaction tasks using the Coordinator and native engine. This is the default behavior for Druid.

> ## Overlord-based

Reviewer: All of the Overlord based stuff should be moved to the end of this page and must be called out as experimental.

You can run automatic compaction using the Overlord rather than the Coordinator. Running compaction tasks on the Overlord means that polling the task status and running compaction at a higher frequency is more efficient than a comparable compaction task that runs on the Coordinator. When running compaction tasks using the Overlord, Druid checks to see if there is data to compact in a datasource every 5 seconds.

* In your Overlord runtime properties, set the following properties:
  *  `druid.supervisor.compaction.enabled` to `true` so that compaction tasks can be run as a supervisor task
  *  `druid.supervisor.compaction.defaultEngine` to  `msq` to specify the MSQ task engine as the compaction engine or to `native`.

After making these changes, you can submit automatic compaction tasks as supervisors. For more general information about supervisors, see [Supervisors](../ingestion/supervisor.md).

## Automatic compaction syntax

You can configure automatic compaction dynamically without restarting Druid.

> Both the native and MSQ task engine automatic compaction engines use the following syntax:

Reviewer: These passages probably don't need to refer to Overlord-based compaction and/or the MSQ engine. For the most part, we should leave the current docs untouched and just add a section about Overlord-based compaction at the end and call out the similarities / differences there.

Reviewer 2: Agree. Since there is no change to the existing flow and everything newly added is currently experimental, we should have it all separated out to the end of the doc.

    "granularitySpec": <compaction task granularitySpec>,
    "skipOffsetFromLatest": <time period to avoid compaction>,
    "taskPriority": <compaction task priority>,
    "taskContext": <task context>,

>     "engine": <native|msq>

Reviewer 2: `engine` can be removed from here since only native is supported with coordinator-driven compaction.

```

## Use Overlord-based automatic compaction

When you use the Overlord for automatic compaction, Druid uses a supervisor task on the Overlord to perform the compaction. Since it's a supervisor task, automatic compaction using the Overlord can run frequently while providing faster compaction times as well as better memory tuning and usage. 

When you use Overlord-based automatic compaction, you can use either the native engine like Coordinator-based automatic compaction or the [MSQ task engine](#use-msq-for-automatic-compaction). 

By default, Druid checks every 5 seconds to see whether or not compaction is required. 

### Use MSQ for automatic compaction

The MSQ task engine is available as a compaction engine if you configure compaction tasks to run on the Overlord as a supervisor. To use the MSQ task engine for automatic compaction, make sure the following requirements are met:

* Have the [MSQ  task engine extension loaded](../multi-stage-query/index.md#load-the-extension).
* In your Overlord runtime properties, set the following properties:
  *  `druid.supervisor.compaction.enabled` to `true` so that compaction tasks can be run as a supervisor task

>   *  `druid.supervisor.compaction.defaultEngine` to `msq` to specify the MSQ task engine as the compaction engine

Reviewer 2: `defaultEngine` is updated to `engine` now
```suggestion
  *  `druid.supervisor.compaction.engine` to `msq` to specify the MSQ task engine as the compaction engine
```

### API for Overlord-based compaction

Submitting an automatic compaction as a supervisor task uses the same endpoint as supervisor tasks for streaming ingestion.

The following example configures auto-compaction for the `wikipedia` datasource:

```sh
curl --location --request POST 'http://localhost:8081/druid/indexer/v1/supervisor' \
--header 'Content-Type: application/json' \
--data-raw '{
   "type": "autocompact",    // required
   "suspended": false,         // optional
   "spec": {                           // required
       "dataSource": "wikipedia",          // required
       "tuningConfig": {...},                    // optional
       "granularitySpec": {...},               // optional
       ...
   }
}'
```

To stop the automatic compaction task, suspend or terminate the supervisor through the UI or API.

### MSQ task engine limitations

When using the MSQ task engine for auto-compaction, keep the following limitations in mind:

- The `metricSpec` field is only supported for idempotent aggregators. For more information, see [Idempotent aggregators](#idempotent-aggregators).
- Only dynamic and range-based partitioning are supported

> - Set `rollup`  to `true` if `metricSpec` is not empty or null. If `metricSpec` is empty or null, set `rollup` to `false`.

Reviewer 2: If metrics-spec is empty or null, a user can set rollup as `false` or null, so we can omit it that part.
```suggestion
- Set `rollup`  to `true` if and only if `metricSpec` is not empty or null.
```

> - You cannot group on multi-value dimensions

Reviewer 2: Limitation of group-by on MVD is removed now. Updating to a different limitation.
```suggestion
- You can only partition on string dimensions. However, multi-valued string dimensions are not supported.
```

- The `metricSpec` field is only supported for idempotent aggregators. For more information, see [Idempotent aggregators](#idempotent-aggregators).
- Only dynamic and range-based partitioning are supported
- Set `rollup`  to `true` if `metricSpec` is not empty or null. If `metricSpec` is empty or null, set `rollup` to `false`.
- You cannot group on multi-value dimensions
- The `maxTotalRows` config is not supported in `DynamicPartitionsSpec`. Use `maxRowsPerSegment` instead.

#### Idempotent aggregators

Idempotent aggregators are aggregators that can be applied repeatedly on a column and each run produces the same results, such as the following `longSum` aggregator:

```
{"name": "added", "type": "longSum", "fieldName": "added"}
```

where the input and output column are both `added`.

The following are some examples of non-idempotent aggregators where each run of the aggregator produces different results:

*  `longSum` aggregator where the `added` column rolls up into the `sum_added` column:
    ```
    {"name": "sum_added", "type": "longSum", "fieldName": "added" }
    ```
* Partial sketches:
    ```

>     {"name": added, "type":"", fieldName: added}

Author: Oops, not intentional. Should it be `HLLSketchMerge`?

>     {"name": added, "type":"", fieldName: added}

Reviewer 2: yes

1. Click **...** > **Submit JSON supervisor**.
1. In the dialog, include the following:
     - The type of supervisor spec by setting `"type": "autocompact"`
     - The compaction configuration by adding it to the `spec` field
    ```json
    {
   "type": "autocompact",
   "spec": {
      "dataSource": YOUR_DATASOURCE,
    ...
    ...
   }
    ```
1. Submit the supervisor.

To stop the automatic compaction task, suspend or terminate the supervisor through the UI or API.

### Manage compaction supervisors with supervisor APIs

Submitting an automatic compaction as a supervisor task uses the same endpoint as supervisor tasks for streaming ingestion.

The following example configures auto-compaction for the `wikipedia` datasource:

```sh
curl --location --request POST 'http://localhost:8081/druid/indexer/v1/supervisor' \
--header 'Content-Type: application/json' \
--data-raw '{
   "type": "autocompact",    // required
   "suspended": false,         // optional
   "spec": {                           // required
       "dataSource": "wikipedia",          // required
       "tuningConfig": {...},                    // optional
       "granularitySpec": {...},               // optional

>        ...

Reviewer 2: I understand that skipping it here and specifying it later in supervisor-based spec may be confusing. If we keep it here, just want to make sure that users realize that it's only supported with supervisors.

Also, we need to add this field to `Automatic compaction dynamic configuration` page. Maybe this info can reside there simiar to below:

`engine `| Engine for compaction. Can be either `native` or `msq`. MSQ is only supported with compaction supervisors <hyperlinked to the section on this page> | no (default = `native`)

>        ...

Reviewer 2: Shall we also include the above on `Automatic compaction dynamic configuration` page?

       "dataSource": "wikipedia",          // required
       "tuningConfig": {...},                    // optional
       "granularitySpec": {...},               // optional
       "engine": <native|msq>,  //optional
       ...
   }
}'
```

Note that if you omit `spec.engine`, Druid uses the default compaction engine. You can control the default compaction engine with the `druid.supervisor.compaction.engine` Overlord runtime property. If `spec.engine` and `druid.supervisor.compaction.engine` are omitted, Druid defaults to the native engine.

To stop the automatic compaction task, suspend or terminate the supervisor through the UI or API.

### Use MSQ for auto-compaction

The MSQ task engine is available as a compaction engine if you configure auto-compaction to use compaction supervisors. To use the MSQ task engine for automatic compaction, make sure the following requirements are met:

* Have the [MSQ  task engine extension loaded](../multi-stage-query/index.md#load-the-extension).
* In your Overlord runtime properties, set the following properties:
  *  `druid.supervisor.compaction.enabled` to `true` so that compaction tasks can be run as a supervisor task

>   *  Optionally, set `druid.supervisor.compaction.engine` to `msq` to specify the MSQ task engine as the default compaction engine. If you don't do this, you'll need to set `spec.engine` to `msq` for each compaction supervisor spec where you want to use the MSQ task engine.

Reviewer 3: Setting this also in Overlord?

To submit an automatic compaction task, you submit a supervisor spec through the UI or API with the type `autocompact` and the `spec` where you define the compaction behavior using the [automatic compaction syntax](#auto-compaction-syntax). You can also use the [web console](#manage-compaction-supervisors-with-the-web-console).

#### MSQ task engine limitations

When using the MSQ task engine for auto-compaction, keep the following limitations in mind:

- The `metricSpec` field is only supported for certain aggregators. For more information, see [Supported aggregators](#supported-aggregators).
- Only dynamic and range-based partitioning are supported
- Set `rollup`  to `true` if and only if `metricSpec` is not empty or null.
- You can only partition on string dimensions. However, multi-valued string dimensions are not supported.

> - The `maxTotalRows` config is not supported in `DynamicPartitionsSpec`. Use `maxRowsPerSegment` instead.

Reviewer 2: Segment sorting limitation is lifted now but holds true for commits in druid 31
```suggestion
- The `maxTotalRows` config is not supported in `DynamicPartitionsSpec`. Use `maxRowsPerSegment` instead.
- Segments can only be sorted on `__time` as the first column.
```

> ##### Idempotent aggregators

Reviewer 2: Wanted to update this section based on this discussion: https://github.com/apache/druid/pull/17250#discussion_r1788835338

To stop the automatic compaction task, suspend or terminate the supervisor through the UI or API.

### Use MSQ for auto-compaction

The MSQ task engine is available as a compaction engine if you configure auto-compaction to use compaction supervisors. To use the MSQ task engine for automatic compaction, make sure the following requirements are met:

* Have the [MSQ  task engine extension loaded](../multi-stage-query/index.md#load-the-extension).
* In your Overlord runtime properties, set the following properties:
  *  `druid.supervisor.compaction.enabled` to `true` so that compaction tasks can be run as a supervisor task
  *  Optionally, set `druid.supervisor.compaction.engine` to `msq` to specify the MSQ task engine as the default compaction engine. If you don't do this, you'll need to set `spec.engine` to `msq` for each compaction supervisor spec where you want to use the MSQ task engine.
* Have at least two compaction task slots available or set `compactionConfig.taskContext.maxNumTasks` to two or more. The MSQ task engine requires at least two tasks to run, one controller task and one worker task.

> You can use [MSQ task engine context parameters](../multi-stage-query/) in `compactionConfig.taskContext` when configuring your datasource for automatic compaction, such as setting the maximum number of tasks using the `compactionConfig.taskContext.maxNumTasks` parameter. Some of the MSQ task engine context parameters overlap with automatic compaction parameters. When these settings overlap, set one or the other.

Reviewer 2: ```suggestion
You can use [MSQ task engine context parameters](../multi-stage-query/) in `spec.taskContext` when configuring your datasource for automatic compaction, such as setting the maximum number of tasks using the `spec.taskContext.maxNumTasks` parameter. Some of the MSQ task engine context parameters overlap with automatic compaction parameters. When these settings overlap, set one or the other.
```

Reviewer 2: I'm not sure if `MSQ task engine context` hyperlink works. Maybe it would once the PR is merged. Please verify once.

}
```

## Auto-compaction using compaction supervisors  

:::info Experimental
Compaction supervisors are experimental. For production use, we recommend [Coordinator-based auto-compaction](#auto-compaction-using-coordinator-duties).
:::

You can run automatic compaction using compaction supervisors on the Overlord rather than Coordinator duties. Compaction supervisors provide the following benefits over Coordinator duties:

* Can use the supervisor framework to get information about the auto-compaction, such as status or state
* More easily suspend or resume compaction for a datasource
* Can use either the native compaction engine or the [MSQ task engine](#use-msq-for-auto-compaction)

> * More reactive and submits tasks as soon as a compaction slot is available

Author: That can come once the feature is more developed

* Tracked compaction task status to avoid re-compacting an interval repeatedly

To use compaction supervisors, set the following properties in your Overlord runtime properties:
  *  `druid.supervisor.compaction.enabled` to `true` so that compaction tasks can be run as a supervisor task
  *  `druid.supervisor.compaction.engine` to  `msq` to specify the MSQ task engine as the compaction engine or to `native` to use the native engine.

Compaction uses the same syntax as Coordinator-based auto-compaction with some differences. Specifically, you submit a supervisor spec with the `type` set to `autocompact` and the auto-compaction config in the `spec` to configure auto-compaction.
  
For information about the syntax, see [automatic-compaction syntax](#auto-compaction-syntax). 

### Manage compaction supervisors with the web console

To submit a supervisor spec for MSQ task engine automatic compaction, perform the following steps:

1. In the web console, go to the **Supervisors** tab.
1. Click **...** > **Submit JSON supervisor**.
1. In the dialog, include the following:
     - The type of supervisor spec by setting `"type": "autocompact"`

>      - The compaction configuration by adding it to the `spec` field

Reviewer 4: Should this be included in the code block?

```sh
curl --location --request POST 'http://localhost:8081/druid/indexer/v1/supervisor' \
--header 'Content-Type: application/json' \
--data-raw '{
   "type": "autocompact",    // required
   "suspended": false,         // optional
   "spec": {                           // required
       "dataSource": "wikipedia",          // required
       "tuningConfig": {...},                    // optional
       "granularitySpec": {...},               // optional
       "engine": <native|msq>,  //optional
       ...
   }
}'
```

Note that if you omit `spec.engine`, Druid uses the default compaction engine. You can control the default compaction engine with the `druid.supervisor.compaction.engine` Overlord runtime property. If `spec.engine` and `druid.supervisor.compaction.engine` are omitted, Druid defaults to the native engine.

To stop the automatic compaction task, suspend or terminate the supervisor through the UI or API.

### Use MSQ for auto-compaction

The MSQ task engine is available as a compaction engine if you configure auto-compaction to use compaction supervisors. To use the MSQ task engine for automatic compaction, make sure the following requirements are met:

> * Have the [MSQ  task engine extension loaded](../multi-stage-query/index.md#load-the-extension).

Reviewer 4: ```suggestion
* [Load the MSQ task engine extension](../multi-stage-query/index.md#load-the-extension).
```

* In your Overlord runtime properties, set the following properties:

>   *  `druid.supervisor.compaction.enabled` to `true` so that compaction tasks can be run as a supervisor task

Reviewer 4: ```suggestion
  *  `druid.supervisor.compaction.enabled` to `true` so that compaction tasks can be run as a supervisor task.
```

  *  Optionally, set `druid.supervisor.compaction.engine` to `msq` to specify the MSQ task engine as the default compaction engine. If you don't do this, you'll need to set `spec.engine` to `msq` for each compaction supervisor spec where you want to use the MSQ task engine.
* Have at least two compaction task slots available or set `compactionConfig.taskContext.maxNumTasks` to two or more. The MSQ task engine requires at least two tasks to run, one controller task and one worker task.

You can use [MSQ task engine context parameters](../multi-stage-query/reference.md#context-parameters) in `spec.taskContext` when configuring your datasource for automatic compaction, such as setting the maximum number of tasks using the `spec.taskContext.maxNumTasks` parameter. Some of the MSQ task engine context parameters overlap with automatic compaction parameters. When these settings overlap, set one or the other.

> To submit an automatic compaction task, you submit a supervisor spec through the UI or API with the type `autocompact` and the `spec` where you define the compaction behavior using the [automatic compaction syntax](#auto-compaction-syntax). You can also use the [web console](#manage-compaction-supervisors-with-the-web-console).

Reviewer 4: >To submit an automatic compaction task, you submit a supervisor spec through the UI or API with the type `autocompact` and the `spec` where you define the compaction behavior using the [automatic compaction syntax](#auto-compaction-syntax). You can also use the [web console](#manage-compaction-supervisors-with-the-web-console).

Is this redundant with the two previous sections?

#### MSQ task engine limitations

When using the MSQ task engine for auto-compaction, keep the following limitations in mind:

- The `metricSpec` field is only supported for certain aggregators. For more information, see [Supported aggregators](#supported-aggregators).

> - Only dynamic and range-based partitioning are supported

Reviewer 4: ```suggestion
- Only dynamic and range-based partitioning are supported.
```

- Set `rollup`  to `true` if and only if `metricSpec` is not empty or null.
- You can only partition on string dimensions. However, multi-valued string dimensions are not supported.
- The `maxTotalRows` config is not supported in `DynamicPartitionsSpec`. Use `maxRowsPerSegment` instead.
- Segments can only be sorted on `__time` as the first column.

##### Supported aggregators

Auto-compaction using the MSQ task engine supports only aggregators that satisfy the following properties: 

> a) mergeability: can also be used to combine partial aggregates
> b) idempotency: produce the same results on repeated runs of the aggregator on previously aggregated values in a column

Reviewer 4: ```suggestion
* __Mergeability__: can combine partial aggregates
* __Idempotency__: produces the same results on repeated runs of the aggregator on previously aggregated values in a column
```

This is exemplified by the following `longSum` aggregator:

```
{"name": "added", "type": "longSum", "fieldName": "added"}
```

where `longSum` being capable of combining partial results satisfies mergeability, while input and output column being the same (`added`) ensures idempotency.

> The following are some examples of aggregators that aren't supported since at least of the required conditions aren't satisfied:

Reviewer 2: ```suggestion
The following are some examples of aggregators that aren't supported since at least one of the required conditions aren't satisfied:
```

*  `longSum` aggregator where the `added` column rolls up into `sum_added` column discarding the input `added` column, violating idempotency, as subsequent runs would no longer find the `added` column:
    ```

>     {"name": "sum_added", "type": "longSum", "fieldName": "added" }

Reviewer 4: ```suggestion
    {"name": "sum_added", "type": "longSum", "fieldName": "added"}
```

    ```
* Partial sketches which cannot themselves be used to combine partial aggregates and need merging aggregators -- such as `HLLSketchMerge` required for `HLLSketchBuild` aggregator below -- violating mergeability:
    ```

>     {"name": added, "type":"HLLSketchBuild", fieldName: added}

Reviewer 4: ```suggestion
    {"name": added, "type":"HLLSketchBuild", fieldName: "added"}
```

    ```
* Count aggregator since it cannot be used to combine partial aggregates and it rolls up into a different `count` column discarding the input column(s), violating both mergeability and idempotency.
    ```

>     { "type" : "count", "name" : "count" }

Reviewer 4: ```suggestion
    {"type": "count", "name": "count"}
```

}
```

## Overlord-based auto-compaction

:::info Experimental
Overlord-based auto-compaction is experimental. For production use, we recommend [Coordinator-based auto-compaction](#use-coordinator-based-auto-compaction).
:::

> You can run automatic compaction using the Overlord rather than the Coordinator. Running compaction tasks on the Overlord means that auto-compaction runs as a supervisor. Since it's a supervisor task, automatic compaction using the Overlord can run frequently while providing faster compaction times as well as better memory tuning and usage. When running auto-compaction tasks using the Overlord, Druid checks to see if there is data to compact in a datasource every 5 seconds.

Reviewer: We should leave out the `5 seconds` part as it is an implementation detail.

}
```

## Auto-compaction using compaction supervisors  

:::info Experimental
Compaction supervisors are experimental. For production use, we recommend [auto-compaction using Coordinator duties](#auto-compaction-using-coordinator-duties).
:::

You can run automatic compaction using compaction supervisors on the Overlord rather than Coordinator duties. Compaction supervisors provide the following benefits over Coordinator duties:

* Can use the supervisor framework to get information about the auto-compaction, such as status or state
* More easily suspend or resume compaction for a datasource
* Can use either the native compaction engine or the [MSQ task engine](#use-msq-for-auto-compaction)
* More reactive and submits tasks as soon as a compaction slot is available
* Tracked compaction task status to avoid re-compacting an interval repeatedly

To use compaction supervisors, set the following properties in your Overlord runtime properties:

>   *  `druid.supervisor.compaction.enabled` to `true` so that compaction tasks can be run as a supervisor tasks

Reviewer 4: ```suggestion
  *  `druid.supervisor.compaction.enabled` to `true` so that compaction tasks can be run as supervisor tasks
```

  *  `druid.supervisor.compaction.engine` to  `msq` to specify the MSQ task engine as the compaction engine or to `native` to use the native engine. This is the default engine if the `engine` field is omitted from your compaction config

> Compaction supervisors use the same syntax as auto-compaction using  Coordinator duties with one key difference: you submit the auto-compaction as a a supervisor spec. In the spec, set the `type` to `autocompact` and include the auto-compaction config in the `spec` .

Reviewer 4: ```suggestion
Compaction supervisors use the same syntax as auto-compaction using  Coordinator duties with one key difference: you submit the auto-compaction as a a supervisor spec. In the spec, set the `type` to `autocompact` and include the auto-compaction config in the `spec`.
```

To submit an automatic compaction task, you can submit a supervisor spec through the [web console](#manage-compaction-supervisors-with-the-web-console) or the [supervisor API](#manage-compaction-supervisors-with-supervisor-apis).

### Manage compaction supervisors with the web console

To submit a supervisor spec for MSQ task engine automatic compaction, perform the following steps:

1. In the web console, go to the **Supervisors** tab.
1. Click **...** > **Submit JSON supervisor**.
1. In the dialog, include the following:
     - The type of supervisor spec by setting `"type": "autocompact"`
     - The compaction configuration by adding it to the `spec` field

>     ```json
>     {
>    "type": "autocompact",
>    "spec": {
>       "dataSource": YOUR_DATASOURCE,
>       "tuningConfig": {...},
>        "granularitySpec": {...},
>        "engine": <native|msq>,
>        ...
>    }
>     ```

Reviewer 4: ```suggestion
    ```json
    {
     "type": "autocompact",
     "spec": {
       "dataSource": YOUR_DATASOURCE,
       "tuningConfig": {...},
       "granularitySpec": {...},
       "engine": <native|msq>,
       ...
    }
    ```
```

You can use [MSQ task engine context parameters](../multi-stage-query/reference.md#context-parameters) in `spec.taskContext` when configuring your datasource for automatic compaction, such as setting the maximum number of tasks using the `spec.taskContext.maxNumTasks` parameter. Some of the MSQ task engine context parameters overlap with automatic compaction parameters. When these settings overlap, set one or the other.

#### MSQ task engine limitations

<!--This list also exists in multi-stage-query/known-issues-->

When using the MSQ task engine for auto-compaction, keep the following limitations in mind:

- The `metricSpec` field is only supported for certain aggregators. For more information, see [Supported aggregators](#supported-aggregators).
- Only dynamic and range-based partitioning are supported.
- Set `rollup`  to `true` if and only if `metricSpec` is not empty or null.
- You can only partition on string dimensions. However, multi-valued string dimensions are not supported.
- The `maxTotalRows` config is not supported in `DynamicPartitionsSpec`. Use `maxRowsPerSegment` instead.
- Segments can only be sorted on `__time` as the first column.

> ##### Supported aggregators

Reviewer 4: ```suggestion
#### Supported aggregators
```

* __Mergeability__: can combine partial aggregates
* __Idempotency__: produces the same results on repeated runs of the aggregator on previously aggregated values in a column

This is exemplified by the following `longSum` aggregator:

```
{"name": "added", "type": "longSum", "fieldName": "added"}
```

where `longSum` being capable of combining partial results satisfies mergeability, while input and output column being the same (`added`) ensures idempotency.

The following are some examples of aggregators that aren't supported since at least one of the required conditions aren't satisfied:

*  `longSum` aggregator where the `added` column rolls up into `sum_added` column discarding the input `added` column, violating idempotency, as subsequent runs would no longer find the `added` column:
    ```
    {"name": "sum_added", "type": "longSum", "fieldName": "added"}
    ```
* Partial sketches which cannot themselves be used to combine partial aggregates and need merging aggregators -- such as `HLLSketchMerge` required for `HLLSketchBuild` aggregator below -- violating mergeability:
    ```

>     {"name": added, "type":"HLLSketchBuild", fieldName: "added"}

Reviewer 4: ```suggestion
    {"name": "added", "type": "HLLSketchBuild", "fieldName": "added"}
```
